sph/include/sph/find_neighbors.hpp

```diff
@@ -5,13 +5,43 @@
 namespace sph
 {
 
+using cstone::LocalIndex;
+
+template<class T, class KeyType>
+void findNeighborsSph(const T* x, const T* y, const T* z, T* h, LocalIndex firstId, LocalIndex lastId,
+                      const cstone::Box<T>& box, const cstone::OctreeNsView<T, KeyType>& treeView, unsigned ng0,
+                      unsigned ngmax, LocalIndex* neighbors, unsigned* nc)
+{
+    LocalIndex numWork = lastId - firstId;
+
+    unsigned ngmin = ng0 / 4;
+
+#pragma omp parallel for
+    for (LocalIndex i = 0; i < numWork; ++i)
+    {
+        LocalIndex id = i + firstId;
+        nc[i]         = findNeighbors(id, x, y, z, h, treeView, box, ngmax, neighbors + i * ngmax);
+
+        int iteration    = 0;
+        int maxIteration = 10;
+        while (ngmin > nc[i] || nc[i] > ngmax && iteration++ < maxIteration)
+        {
+            h[id] = updateH(ng0, nc[i], h[id]);
+            nc[i] = findNeighbors(id, x, y, z, h, treeView, box, ngmax, neighbors + i * ngmax);
+        }
+    }
```

Comment:
Here's the main algorithmic change on the CPU. On the GPU it's implemented the same.

Reply:
Could we get a warning when particles do not converge? (i.e. when we reach iteration = 10 and still needs more)

Reply:
done